File: LeetCode/73SetMatrixZeros/SetMatrixZeroes.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        // Constant memory -> Time:O(nm), Space:O(1)
        int n = matrix.size();
        int m = matrix[0].size();
        bool firstRow = false;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (matrix[i][j] == 0) {
                    matrix[0][j] = 0;
                    if (i == 0) {
                        firstRow = true;
                    }
                    else {
                        matrix[i][0] = 0;
                    }
                }
            }
        }
        for (int i = 1; i < n; i++) {
            for (int j = 1; j < m; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        if (matrix[0][0] == 0) {
            for (int i = 0; i < n; i++) {
                matrix[i][0] = 0;
            }
        }
        if (firstRow) {
            for (int j = 0; j < m; j++) {
                matrix[0][j] = 0;
            }
        }

        // // 2 vectors -> Time: O(mn), Space:O(m+n)
        // int n = matrix.size();
        // int m = matrix[0].size();
        // vector<int> rows;
        // vector<int> cols;
        // for (int i = 0; i < n; i++) {
        //     for (int j = 0; j < m; j++) {
        //         if (matrix[i][j] == 0) {
        //             rows.push_back(i);
        //             cols.push_back(j);
        //         }
        //     }
        // }
        // for (int i = 0; i < rows.size(); i++) {
        //     for (int j = 0; j < m; j++) {
        //         matrix[rows[i]][j] = 0;
        //     }
        // }
        // for (int j = 0; j < cols.size(); j++) {
        //     for (int i = 0; i < n; i++) {
        //         matrix[i][cols[j]] = 0;
        //     }
        // }
    }
};

File: LeetCode/73SetMatrixZeros/SetMatrixZeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # Constant space -> Time:O(nm), Space:O(1)
        first_row = False
        n = len(matrix)
        m = len(matrix[0])
        for i in range(n):
            for j in range(m):
                if matrix[i][j] == 0:
                    if i == 0:
                        first_row = True
                    else:
                        matrix[i][0] = 0
                    matrix[0][j] = 0
        for i in range(1, n):
            for j in range(1, m):
                if matrix[0][j] == 0 or matrix[i][0] == 0:
                    matrix[i][j] = 0
        if matrix[0][0] == 0:
            for i in range(n):
                matrix[i][0] = 0
        if first_row:
            for j in range(m):
                matrix[0][j] = 0

        # # rows and cols array -> Time:O(nm), Space:O(n+m)
        # rows = []
        # cols = []
        # n = len(matrix)
        # m = len(matrix[0])
        # for i in range(n):
        #     for j in range(m):
        #         if matrix[i][j] == 0:
        #             rows.append(i)
        #             cols.append(j)
        # for i in rows:
        #     for j in range(m):
        #         matrix[i][j] = 0
        # for j in cols:
        #     for i in range(n):
        #         matrix[i][j] = 0
